File: src/physics/shapecast.ts
import RAPIER from '@dimforge/rapier3d-compat';
import type { Vector3 } from 'three';

/**
 * Collision *queries only* — see docs/source-movement.md. The accel/friction
 * curve and the collide-and-slide loop live in src/player/movement.ts by hand;
 * this module just wraps Rapier's shape-cast so movement.ts doesn't touch the
 * Rapier API directly.
 */

// Capsules here are never rotated (always upright), so a shared identity
// quaternion avoids allocating one per cast.
const IDENTITY_ROTATION: RAPIER.Rotation = { x: 0, y: 0, z: 0, w: 1 };

// If the swept shape ends up closer than this to a surface, Rapier reports a
// hit rather than requiring exact penetration — cheap numerical safety margin
// for a character controller. Not a ported Source value.
const TARGET_DISTANCE = 0.005;

/**
 * Sweeps a vertical capsule from `origin` (capsule centre) by `displacement`
 * against everything in `world` except `excludeCollider`. Returns the impact
 * fraction in [0, 1] along `displacement`, or null on no hit, and writes the
 * world-space hit normal into `outNormal`. No allocation beyond what Rapier's
 * own castShape() call unavoidably does internally.
 */
export function capsuleCast(
  world: RAPIER.World,
  shape: RAPIER.Capsule,
  origin: Vector3,
  displacement: Vector3,
  outNormal: Vector3,
  excludeCollider?: RAPIER.Collider,
): number | null {
  if (displacement.lengthSq() === 0) return null;

  const hit = world.castShape(
    origin,
    IDENTITY_ROTATION,
    displacement,
    shape,
    TARGET_DISTANCE,
    1.0,
    true,
    undefined,
    undefined,
    excludeCollider,
  );
  if (!hit) return null;

  outNormal.set(hit.normal1.x, hit.normal1.y, hit.normal1.z);
  return hit.time_of_impact;
}

// Reused across shots — a bullet trace runs at most once per tick, and the Ray
// fields are plain vectors we overwrite.
const scratchRay = new RAPIER.Ray({ x: 0, y: 0, z: 0 }, { x: 0, y: 0, z: 0 });

/**
 * Traces a bullet: casts a ray from `origin` along the unit vector `direction`,
 * up to `maxDistance` metres. Returns the distance to the impact and writes the
 * surface normal into `outNormal`, or null if the ray hits nothing. The hit
 * point is `origin + direction * distance`.
 *
 * This is the world half of the shot pipeline — the per-bone hitbox query
 * against a character rig is separate and still owed (needs the Phase 3 rig).
 */
export function rayCast(
  world: RAPIER.World,
  origin: Vector3,
  direction: Vector3,
  maxDistance: number,
  outNormal: Vector3,
  excludeCollider?: RAPIER.Collider,
): number | null {
  scratchRay.origin = origin;
  scratchRay.dir = direction;
  // solid=true: a ray starting inside a collider impacts at distance 0 rather
  // than passing out through the far face.
  const hit = world.castRayAndGetNormal(scratchRay, maxDistance, true, undefined, undefined, excludeCollider);
  if (!hit) return null;

  outNormal.set(hit.normal.x, hit.normal.y, hit.normal.z);
  return hit.timeOfImpact; // direction is unit, so toi is metres
}

/** True if a capsule at `center` overlaps anything in `world` (except `excludeCollider`).
 * Used for the "can I stand up" headroom check when un-ducking — see handleDuck(). */
export function capsuleOverlapsAnything(
  world: RAPIER.World,
  shape: RAPIER.Capsule,
  center: Vector3,
  excludeCollider?: RAPIER.Collider,
): boolean {
  let found = false;
  world.intersectionsWithShape(
    center,
    IDENTITY_ROTATION,
    shape,
    () => {
      found = true;
      return false; // stop at first hit
    },
    undefined,
    undefined,
    excludeCollider,
  );
  return found;
}

File: sim/src/shapecast.rs
/// Shapecast wrappers for Rapier3D — the only place Rapier is used for collision
/// queries. See docs/source-movement.md: the accel/friction curve and the
/// collide-and-slide loop live in movement.rs by hand; this module just wraps
/// Rapier's shape-cast/ray-cast/overlap queries.
///
/// All coordinates use f64 for consistency with the sim's movement math;
/// conversion to f32 for Rapier happens at this boundary.
use rapier3d::parry::query::details::ShapeCastOptions;
use rapier3d::prelude::*;

/// If a swept shape ends up closer than this to a surface, Rapier reports a hit
/// rather than requiring exact penetration. Not a ported Source value.
const TARGET_DISTANCE: f32 = 0.005;

/// Sweeps a shape from `origin` (world-space) by `displacement` metres against
/// everything in `physics` except `exclude_collider`. Returns the impact fraction
/// in [0, 1], or None on no hit, and writes the world-space hit normal into
/// `out_normal`.
pub fn capsule_cast(
    physics: &PhysicsWorld,
    shape: &dyn Shape,
    origin_x: f64, origin_y: f64, origin_z: f64,
    disp_x: f64, disp_y: f64, disp_z: f64,
    out_normal: &mut nalgebra::Vector3<f64>,
    exclude_collider: ColliderHandle,
    stop_at_penetration: bool,
) -> Option<f64> {
    let disp_sq = (disp_x * disp_x + disp_y * disp_y + disp_z * disp_z) as f32;
    if disp_sq == 0.0 {
        return None;
    }

    let shape_pos = Pose::from_parts(
        Vector::new(origin_x as f32, origin_y as f32, origin_z as f32),
        Rotation::IDENTITY,
    );
    let shape_vel = Vector::new(disp_x as f32, disp_y as f32, disp_z as f32);

    let options = ShapeCastOptions {
        target_distance: TARGET_DISTANCE,
        stop_at_penetration,
        max_time_of_impact: 1.0,
        compute_impact_geometry_on_penetration: true,
    };

    let filter = QueryFilter::default().exclude_collider(exclude_collider);

    let hit = physics.cast_shape(&shape_pos, shape_vel, shape, options, filter);
    match hit {
        Some((_collider_handle, shape_cast_hit)) => {
            out_normal.x = shape_cast_hit.normal1.x as f64;
            out_normal.y = shape_cast_hit.normal1.y as f64;
            out_normal.z = shape_cast_hit.normal1.z as f64;
            Some(shape_cast_hit.time_of_impact as f64)
        }
        None => None,
    }
}

/// True if a shape at `center` overlaps anything in `physics` (except `exclude_collider`).
pub fn capsule_overlaps_anything(
    physics: &PhysicsWorld,
    shape: &dyn Shape,
    center_x: f64, center_y: f64, center_z: f64,
    exclude_collider: ColliderHandle,
) -> bool {
    let shape_pos = Pose::from_parts(
        Vector::new(center_x as f32, center_y as f32, center_z as f32),
        Rotation::IDENTITY,
    );
    let filter = QueryFilter::default().exclude_collider(exclude_collider);

    physics.intersect_shape(shape_pos, shape, filter).next().is_some()
}

/// Traces a ray from `origin` along the unit vector `direction`, up to `max_distance`
/// metres. Returns the distance to the impact and writes the surface normal into
/// `out_normal`, or None if the ray hits nothing.
pub fn ray_cast(
    physics: &PhysicsWorld,
    origin_x: f64, origin_y: f64, origin_z: f64,
    dir_x: f64, dir_y: f64, dir_z: f64,
    max_distance: f64,
    out_normal: &mut nalgebra::Vector3<f64>,
    exclude_collider: ColliderHandle,
) -> Option<f64> {
    let ray = Ray::new(
        Vector::new(origin_x as f32, origin_y as f32, origin_z as f32),
        Vector::new(dir_x as f32, dir_y as f32, dir_z as f32),
    );
    let filter = QueryFilter::default().exclude_collider(exclude_collider);
    let hit = physics.cast_ray_and_get_normal(&ray, max_distance as f32, true, filter);
    match hit {
        Some((_collider_handle, intersection)) => {
            out_normal.x = intersection.normal.x as f64;
            out_normal.y = intersection.normal.y as f64;
            out_normal.z = intersection.normal.z as f64;
            Some(intersection.time_of_impact as f64)
        }
        None => None,
    }
}
